### PR TITLE
DTS-3385: migrate getTableVersion to use delta-kernel 

### PR DESCRIPTION
**Pull request by pranavsuku-db**
getTableVersion now uses the DeltaSharedTableKernel implementation. In this PR, delta-kernel is imported into OSS and DeltaSharedTableKernel and the original DeltaSharedTable inherit from a trait DeltaSharedTableProtocol. This protocol trait is what is returned by the table loader. The table loader will return a standalone DeltaSharedTable unless a parameter is specified to return the kernel implementation. This parameter is only turned on for getTableVersion.

### CODE REVIEW

File: server/src/main/scala/io/delta/kernelsharedtable/DeltaSharedTableKernel.scala [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+                                       jsonPredicateHints: Option[String],
+                                       limitHint: Option[Long],
+                                       includeHistoricalMetadata: Option[Boolean])
+
+
+/**
+ * A table class that wraps `DeltaLog` to provide the methods used by the server.
+ */
+class DeltaSharedTableKernel(
+                        tableConfig: TableConfig,
```
Comment (linzhou-db) [inline server/src/main/scala/io/delta/kernelsharedtable/DeltaSharedTableKernel.scala:74]: nit: fix the indent

File: server/src/main/scala/io/delta/kernelsharedtable/DeltaSharedTableKernel.scala [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+                        evaluateJsonPredicateHints: Boolean,
+                        evaluateJsonPredicateHintsV2: Boolean,
+                        queryTablePageSizeLimit: Int,
+                        queryTablePageTokenTtlMs: Int,
+                        refreshTokenTtlMs: Int) extends DeltaSharedTableProtocol {
+
+  protected val tablePath: Path = new Path(tableConfig.getLocation)
+
+
+  // Get the table and table client (engine). If shouldParallelLoadDeltaLog is true,
```
Comment (linzhou-db) [inline server/src/main/scala/io/delta/kernelsharedtable/DeltaSharedTableKernel.scala:86]: nit: update the comment, shouldParallelLoadDeltaLog is not available, remove related content.

File: server/src/main/scala/io/delta/kernelsharedtable/DeltaSharedTableKernel.scala [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+      val (table, engine) = getTableAndEngine()
+      val snapshot = table.getLatestSnapshot(engine)
+      snapshot.getVersion(engine)
+
+    }
+  }
+
+  // scalastyle:off argcount
+  override def query(
+             includeFiles: Boolean,
```
Comment (linzhou-db) [inline server/src/main/scala/io/delta/kernelsharedtable/DeltaSharedTableKernel.scala:115]: nit: fix the indent

File: server/src/main/scala/io/delta/kernelsharedtable/DeltaSharedTableKernel.scala [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+             includeRefreshToken: Boolean,
+             refreshToken: Option[String],
+             responseFormatSet: Set[String]): QueryResult = {
+
+    throw new DeltaSharingUnsupportedOperationException("not implemented yet")
+
+  }
+
+  override def queryCDF(
+                         cdfOptions: Map[String, String],
```
Comment (linzhou-db) [inline server/src/main/scala/io/delta/kernelsharedtable/DeltaSharedTableKernel.scala:134]: nit: fix the indent

File: server/src/main/scala/io/delta/sharing/server/DeltaSharedTableProtocol.scala [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+case class QueryResult(
+                        version: Long,
+                        actions: Seq[Object],
+                        responseFormat: String)
+trait DeltaSharedTableProtocol {
+  def getTableVersion(startingTimestamp: Option[String]): Long = -1
+
+  // scalastyle:off argcount
+  def query(
+             includeFiles: Boolean,
```
Comment (linzhou-db) [inline server/src/main/scala/io/delta/sharing/server/DeltaSharedTableProtocol.scala:32]: nit: fix the indent

File: server/src/main/scala/io/delta/sharing/server/DeltaSharedTableProtocol.scala [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+             startingVersion: Option[Long],
+             endingVersion: Option[Long],
+             maxFiles: Option[Int],
+             pageToken: Option[String],
+             includeRefreshToken: Boolean,
+             refreshToken: Option[String],
+             responseFormatSet: Set[String]): QueryResult
+
+  def queryCDF(
+                cdfOptions: Map[String, String],
```
Comment (linzhou-db) [inline server/src/main/scala/io/delta/sharing/server/DeltaSharedTableProtocol.scala:47]: nit: fix the indent

File: server/src/main/scala/io/delta/kernelsharedtable/DeltaSharedTableKernel.scala [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+import io.delta.sharing.server.protocol.{QueryTablePageToken, RefreshToken}
+import io.delta.sharing.server.util.JsonUtils
+
+
+/**
+ * A util class stores all query parameters. Used to compute the checksum in the page token for
+ * query validation.
+ */
+private case class QueryParamChecksum(
+                                       version: Option[Long],
```
[the remark this answers is not in the document]
Comment (pranavsuku-db) [reply server/src/main/scala/io/delta/kernelsharedtable/DeltaSharedTableKernel.scala:58]: fixed

File: server/src/main/scala/io/delta/kernelsharedtable/DeltaSharedTableKernel.scala (hunk 1 above)
Comment (pranavsuku-db) [reply server/src/main/scala/io/delta/kernelsharedtable/DeltaSharedTableKernel.scala:74]: fixed

File: server/src/main/scala/io/delta/kernelsharedtable/DeltaSharedTableKernel.scala (hunk 2 above)
Comment (pranavsuku-db) [reply server/src/main/scala/io/delta/kernelsharedtable/DeltaSharedTableKernel.scala:86]: fixed

File: server/src/main/scala/io/delta/kernelsharedtable/DeltaSharedTableKernel.scala (hunk 3 above)
Comment (pranavsuku-db) [reply server/src/main/scala/io/delta/kernelsharedtable/DeltaSharedTableKernel.scala:115]: fixed

File: server/src/main/scala/io/delta/kernelsharedtable/DeltaSharedTableKernel.scala (hunk 4 above)
Comment (pranavsuku-db) [reply server/src/main/scala/io/delta/kernelsharedtable/DeltaSharedTableKernel.scala:134]: fixed

File: server/src/main/scala/io/delta/sharing/server/DeltaSharedTableProtocol.scala (hunk 5 above)
Comment (pranavsuku-db) [reply server/src/main/scala/io/delta/sharing/server/DeltaSharedTableProtocol.scala:32]: fixed

File: server/src/main/scala/io/delta/sharing/server/DeltaSharedTableProtocol.scala (hunk 6 above)
Comment (pranavsuku-db) [reply server/src/main/scala/io/delta/sharing/server/DeltaSharedTableProtocol.scala:47]: fixed

File: server/src/main/scala/io/delta/kernelsharedtable/DeltaSharedTableKernel.scala [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+   *  the latest table version.
+   */
+  override def getTableVersion(startingTimestamp: Option[String]): Long = {
+    if (startingTimestamp.isDefined) {
+      throw new DeltaSharingIllegalArgumentException("starting timestamp not defined")
+    } else {
+      val (table, engine) = getTableAndEngine()
+      val snapshot = table.getLatestSnapshot(engine)
+      snapshot.getVersion(engine)
+
```
Comment (linzhou-db) [inline server/src/main/scala/io/delta/kernelsharedtable/DeltaSharedTableKernel.scala:108]: nit: remove empty line?

**Review by linzhou-db — APPROVED** [review]

File: server/src/main/scala/io/delta/kernelsharedtable/DeltaSharedTableKernel.scala (hunk 8 above)
Comment (pranavsuku-db) [reply server/src/main/scala/io/delta/kernelsharedtable/DeltaSharedTableKernel.scala:108]: will fix in next pr

**Pull request merged**